src/main/java/io/jenkins/plugins/httpclient/RobustHTTPClient.java

```diff
@@ -0,0 +1,258 @@
+     * Number of upload/download attempts of nonfatal errors before giving up.
+     */
+    public void setStopAfterAttemptNumber(int stopAfterAttemptNumber) {
+        this.stopAfterAttemptNumber = stopAfterAttemptNumber;
+    }
+
+    /**
+     * Initial number of milliseconds between first and second upload/download attempts.
+     * Subsequent ones increase exponentially.
+     * Note that this is not a <em>randomized</em> exponential backoff;
+     * and the base of the exponent is currently hard-coded to 2.
+     */
+    public void setWaitMultiplier(long waitMultiplier) {
+        this.waitMultiplier = waitMultiplier;
+    }
+
+    /**
+     * Maximum number of seconds between upload/download attempts.
+     */
+    public void setWaitMaximum(long waitMaximum) {
+        this.waitMaximum = waitMaximum;
+    }
+
+    /**
+     * Number of seconds to permit a single upload/download attempt to take.
+     */
+    public void setTimeout(long timeout) {
+        this.timeout = timeout;
+    }
+
+    @FunctionalInterface
+    public interface ConnectionCreator {
+        CloseableHttpResponse connect(CloseableHttpClient client) throws IOException, InterruptedException;
+    }
+
+    @FunctionalInterface
+    public interface ConnectionUser {
+        void use(CloseableHttpResponse response) throws IOException, InterruptedException;
+    }
+
+    /**
+     * Perform an HTTP network operation with appropriate timeouts and retries.
+     * @param whatConcise a short description of the operation, like {@code upload}
+     * @param whatVerbose a longer description of the operation, like {@code uploading … to …}
+     * @param connectionCreator how to establish a connection prior to getting the server’s response
+     * @param connectionUser what to do, if anything, after a successful (2xx) server response
+     * @param listener a place to print messages
+     * @throws IOException if there is an unrecoverable error; {@link AbortException} will be used where appropriate
+     * @throws InterruptedException if the transfer is interrupted
+     */
+    public void connect(String whatConcise, String whatVerbose, ConnectionCreator connectionCreator, ConnectionUser connectionUser, TaskListener listener) throws IOException, InterruptedException {
+        AtomicInteger responseCode = new AtomicInteger();
+        int attempt = 1;
+        while (true) {
+            try {
+                try {
+                    executors.submit(() -> {
+                        responseCode.set(0);
+                        try (CloseableHttpClient client = HttpClients.createSystem()) {
+                            try (CloseableHttpResponse response = connectionCreator.connect(client)) {
+                                StatusLine statusLine = response.getStatusLine();
+                                responseCode.set(statusLine != null ? statusLine.getStatusCode() : 0);
+                                if (responseCode.get() < 200 || responseCode.get() >= 300) {
```

Comment:
what haven if there are a redirection 301/302

Reply:
Treated as a failure currently. The existing use cases do not require following redirects. It would be possible to handle I think if there were some use case.